node-api/handlers/beacon/header.go

```diff
@@ -21,58 +21,91 @@
 package beacon
 
 import (
+	"errors"
+	"fmt"
+
 	"github.com/berachain/beacon-kit/node-api/handlers"
 	beacontypes "github.com/berachain/beacon-kit/node-api/handlers/beacon/types"
 	"github.com/berachain/beacon-kit/node-api/handlers/utils"
 	"github.com/berachain/beacon-kit/primitives/math"
 )
 
+var ErrMismatchedSlotAndParentBlock = errors.New("slot does not match with parent block")
+
 func (h *Handler) GetBlockHeaders(c handlers.Context) (any, error) {
-	req, err := utils.BindAndValidate[beacontypes.GetBlockHeadersRequest](
-		c, h.Logger(),
-	)
+	req, err := utils.BindAndValidate[beacontypes.GetBlockHeadersRequest](c, h.Logger())
 	if err != nil {
 		return nil, err
 	}
-	slot, err := math.U64FromString(req.Slot)
-	if err != nil {
-		return nil, err
+
+	if len(req.Slot) == 0 && len(req.ParentRoot) == 0 {
+		// no parameter specified, pick chain HEAD
+		// by requesting special slot 0.
+		return makeBlockHeaderResponse(h.backend, 0, true /*resultsInList*/)
 	}
-	header, err := h.backend.BlockHeaderAtSlot(slot)
-	if err != nil {
-		return nil, err
+
+	var (
+		slot, errSlot         = math.U64FromString(req.Slot)
+		parentSlot, errParent = utils.SlotFromBlockID(req.ParentRoot, h.backend)
+	)
+
+	switch {
+	case errSlot == nil && errParent != nil:
+		return makeBlockHeaderResponse(h.backend, slot, true /*resultsInList*/)
+
+	case errSlot != nil && errParent == nil:
+		return makeBlockHeaderResponse(h.backend, parentSlot+1, true /*resultsInList*/)
+
+	case errSlot == nil && errParent == nil:
+		if slot != parentSlot+1 {
+			return nil, fmt.Errorf("%w: request slot %d, parent block slot %d", ErrMismatchedSlotAndParentBlock, slot, parentSlot)
+		}
+		return makeBlockHeaderResponse(h.backend, slot, true /*resultsInList*/)
+
+	default:
+
+		return nil, fmt.Errorf("failed retrieving slot from input parameters: %w, %w", errSlot, errParent)
 	}
-	return beacontypes.NewResponse(&beacontypes.BlockHeaderResponse{
-		Root:      header.GetBodyRoot(),
-		Canonical: true,
-		Header: &beacontypes.SignedBeaconBlockHeader{
-			Message:   beacontypes.BeaconBlockHeaderFromConsensus(header),
-			Signature: "", // TODO: implement
-		},
-	}), nil
 }
 
 func (h *Handler) GetBlockHeaderByID(c handlers.Context) (any, error) {
-	req, err := utils.BindAndValidate[beacontypes.GetBlockHeaderRequest](
-		c, h.Logger(),
-	)
+	req, err := utils.BindAndValidate[beacontypes.GetBlockHeaderRequest](c, h.Logger())
 	if err != nil {
 		return nil, err
 	}
 	slot, err := utils.SlotFromBlockID(req.BlockID, h.backend)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed retrieving slot from block ID %s: %w", req.BlockID, err)
```

Comment:
unlike prysm, we don't properly set HTTP error code here and it's all a 500 even when it should be a 400.
IIUC this is more of a general issue that this specific PR bug, so I would defer fixing the thing to a separate PR.
CC @calbera, @rezbera